pdns/ws-auth.cc

```diff
@@ -497,27 +520,142 @@ static void gatherComments(const Json container, const DNSName& qname, const QTy
+
+        int k_algo = DNSSECKeeper::shorthand2algorithm(::arg()["default-ksk-algorithm"]);
+        int z_algo = DNSSECKeeper::shorthand2algorithm(::arg()["default-zsk-algorithm"]);
+        int k_size = arg().asNum("default-ksk-size");
+        int z_size = arg().asNum("default-zsk-size");
+
+        if (k_algo != -1) {
+          int64_t id;
+          if (!dk.addKey(zonename, true, k_algo, id, k_size)) {
+            throw ApiException("No backend was able to secure '" + zonename.toString() + "', most likely because no DNSSEC"
+                + "capable backends are loaded, or because the backends have DNSSEC disabled."
+                + "For the Generic SQL backends, set the 'gsqlite3-dnssec', 'gmysql-dnssec' or"
+                + "'gpgsql-dnssec' flag. Also make sure the schema has been updated for DNSSEC!");
+          }
+        }
+
+        if (z_algo != -1) {
+          int64_t id;
+          if (!dk.addKey(zonename, false, z_algo, id, z_size)) {
+            throw ApiException("No backend was able to secure '" + zonename.toString() + "', most likely because no DNSSEC"
+                + "capable backends are loaded, or because the backends have DNSSEC disabled."
+                + "For the Generic SQL backends, set the 'gsqlite3-dnssec', 'gmysql-dnssec' or"
+                + "'gpgsql-dnssec' flag. Also make sure the schema has been updated for DNSSEC!");
+          }
+        }
+
+        // Used later for NSEC3PARAM
+        isDNSSECZone = dk.isSecuredZone(zonename);
+
+        if (!isDNSSECZone) {
+          throw ApiException("Failed to secure '" + zonename.toString() + "'. Is your backend dnssec enabled? (set "
+              + "gsqlite3-dnssec, or gmysql-dnssec etc). Check this first."
+              + "If you run with the BIND backend, make sure you have configured"
+              + "it to use DNSSEC with 'bind-dnssec-db=/path/fname' and"
+              + "'pdnsutil create-bind-db /path/fname'!");
+        }
+        shouldRectify = true;
+      }
+    } else {
+      // "dnssec": false in json
+      if (isDNSSECZone) {
+        throw ApiException("Refusing to un-secure zone " + zonename.toString());
+      }
+    }
+  }
+
+  if(document["nsec3param"].string_value().length() > 0) {
+    shouldRectify = true;
+    NSEC3PARAMRecordContent ns3pr(document["nsec3param"].string_value());
+    string error_msg = "";
+    if (!isDNSSECZone) {
+      throw ApiException("NSEC3PARAMs provided for zone '"+zonename.toString()+"', but zone is not DNSSEC secured.");
+    }
+    if (!dk.checkNSEC3PARAM(ns3pr, error_msg)) {
+      throw ApiException("NSEC3PARAMs provided for zone '"+zonename.toString()+"' are invalid. " + error_msg);
+    }
+    if (!dk.setNSEC3PARAM(zonename, ns3pr, boolFromJson(document, "nsec3narrow", false))) {
+      throw ApiException("NSEC3PARAMs provided for zone '" + zonename.toString() +
+          "' passed our basic sanity checks, but cannot be used with the current backend.");
```

Comment:
s/current/hosting/?

Reply:
I believe current is the better word here; "hosting" over-specifies the backend and suggests there is more than one backend at a time.